### PR TITLE
Debugging format in CI

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
Current random CI failures (#6062) are very cryptic. I think enabling the `--format doc` flag might give some clues?

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [ ] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
Funny, only 8 checks appear in this PR. I think this is because I enabled CircleCI [in my fork](https://circleci.com/gh/deivid-rodriguez/rubocop) to run some experiments, and tripped CircleCI up? :sweat_smile:

**Comment by deivid-rodriguez** [comment]
This is pretty much a WIP, so I'll close it until I have something better!

**Comment by bquorning** [comment]
> Since I started using the documentation format in my fork, I saw tests being killed in the same spec twice

That is very interesting…

Perhaps we could skip that one spec when running JRuby?

**Comment by deivid-rodriguez** [comment]
The problem with jruby seems like a too much memory issue. Spawing jruby subprocesses from jruby seems quite expensive currently in this regard. [This comment](https://github.com/jruby/jruby/issues/3739) exactly matches the failures and the reason, but in our case, tweaking memory limits is not enough.

I actually think I managed to consistently managed to fix this in my fork by turning the `.travis.rb` script into a shell script, since it reduces one level of subshelling _everywhere_. Since `.travis.rb` basically just shells out different processes, the resulting script is simpler IMO. You can see the WIP [here](https://github.com/rubocop-hq/rubocop/compare/master...deivid-rodriguez:debug_ci_fork). I enabled hourly builds of that branch in my fork for a couple of days, and it was consistently green except for [a couple of failures](https://circleci.com/gh/deivid-rodriguez/rubocop/875) that seem not memory related, but a bug in  bundler, I'd say.

**Comment by bquorning** [comment]
So, let me just get this straight: .travis.yml and .circleci/config.yml call `ruby .travis.rb`, which will start a java process. From .travis.rb, we will *three* times in a row call `ruby` or `bundle exec rake`, that all start a 2nd java process. And from the process that runs rspec, we will eventually start a 3rd java process, sometimes (now unsurprisingly) causing CircleCI to run out of memory.

**Comment by deivid-rodriguez** [comment]
Yup.

**Comment by bquorning** [comment]
I think it should be possible to run all that needs running in just *one* process instead of three.

**Comment by deivid-rodriguez** [comment]
I'd say let's start reducing one level the subprocess nesting and then further improve on that if it turns out to be necessary?

File: .ci.rb [hunk 1]
```diff
@@ -16,11 +17,32 @@ def run
     private
 
     # Check requiring libraries successfully.
-    # See https://github.com/rubocop-hq/rubocop/pull/4523#issuecomment-309136113
     def check_requiring_libraries
+      check_require_status
+
+      check_require_output
```
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply .ci.rb]: I like that idea :+1:.

**Comment by bquorning** [comment]
> I actually think I managed to consistently fix this in my fork by turning the .travis.rb script into a shell script, since it reduces one level of subshelling everywhere.

Got it. Do you have that code shared anywhere?

**Comment by deivid-rodriguez** [comment]
Yep, I pointed to it in [this comment](https://github.com/rubocop-hq/rubocop/pull/6203#issuecomment-415987927), it's the branch called `debug_ci_fork` in my fork. Sorry it's a mix of unrelated changes, I'll clean that up when I compile the resulting PRs of all this work.

**Comment by bquorning** [comment]
I found it: https://github.com/rubocop-hq/rubocop/compare/master...deivid-rodriguez:debug_ci_fork

Looking forward to seeing the pull request 😄

**Comment by deivid-rodriguez** [comment]
> I found it: master...deivid-rodriguez:debug_ci_fork

I like how clear github renders diff links by default, I'll start just directly dropping the link in my comments :smile:

> Looking forward to seeing the pull request :smile:

I'll open a few ones so they are focused and easier to review! :+1:

**Comment by bquorning** [comment]
By the way, regarding `if [ -z "$JRUBY_OPTS" ]` you should look at https://circleci.com/docs/2.0/env-vars/#setting-an-environment-variable-in-a-container 🎉

**Comment by deivid-rodriguez** [comment]
Not sure what you mean... :sweat_smile:

**Comment by bquorning** [comment]
Sorry :-) I just meant that with CircleCI config ENV variables can be configures per job, or per container. We can use this to set `JRUBY_OPTS` only on the JRuby jobs, and `TEST_QUEUE_WORKERS` on all jobs with a minimum of duplication. E.g. (with `jruby_env` being the important change):

```yaml
steps: &steps
  steps:
    - checkout
    - run: bundle install
    - run: bash .travis.sh

jruby_env: &jruby_env
  # By default we have 2 CPUs and 4GB ram available
  # https://circleci.com/docs/2.0/configuration-reference/#resource_class
  JRUBY_OPTS: "--debug -J-Xmn1024m -J-Xms2048m -J-Xmx2048m"

common_env: &common_env
  # CircleCI container has two cores, but Ruby can see 32 cores. So we
  # configure test-queue.
  # See https://github.com/tmm1/test-queue#environment-variables
  TEST_QUEUE_WORKERS: 2

rubocop_env: &rubocop_env
  environment:
    TASK: "internal_investigation"
    <<: *common_env

jobs:
  # ... 

  jruby-9.2-rubocop:
    docker:
      - image: circleci/jruby:9.2
        environment:
          <<: *jruby_env
    <<: *rubocop_env
    <<: *steps
```

**Comment by deivid-rodriguez** [comment]
Oh, yeah, sure, I was planning to open a PR with that change too! :)

**Comment by deivid-rodriguez** [comment]
I like your specific proposal a lot, can you PR it @bquorning?

**Comment by bquorning** [comment]
#6225

**Pull request closed without merging**